contracts/integrations/test/function-assertions/moveStake.ts

```diff
@@ -0,0 +1,111 @@
+import {
+    GlobalStakeByStatus,
+    OwnerStakeByStatus,
+    StakeInfo,
+    StakeStatus,
+    StakingPoolById,
+    StoredBalance,
+} from '@0x/contracts-staking';
+import { constants, expect } from '@0x/contracts-test-utils';
+import { BigNumber, logUtils } from '@0x/utils';
+import { TxData } from 'ethereum-types';
+import * as _ from 'lodash';
+
+import { DeploymentManager } from '../utils/deployment_manager';
+import { FunctionAssertion } from '../utils/function_assertions';
+
+function incrementNextEpochBalance(stakeBalance: StoredBalance, amount: BigNumber): void {
+    _.update(stakeBalance, ['nextEpochBalance'], balance => (balance || constants.ZERO_AMOUNT).plus(amount));
+}
+
+function decrementNextEpochBalance(stakeBalance: StoredBalance, amount: BigNumber): void {
+    _.update(stakeBalance, ['nextEpochBalance'], balance => (balance || constants.ZERO_AMOUNT).minus(amount));
+}
+
+/**
+ * Returns a FunctionAssertion for `moveStake` which assumes valid input is provided. The
+ * FunctionAssertion checks that the staker's
+ */
+export function validMoveStakeAssertion(
+    deployment: DeploymentManager,
+    globalStake: GlobalStakeByStatus,
+    ownerStake: OwnerStakeByStatus,
+    pools: StakingPoolById,
+): FunctionAssertion<{}> {
+    const { stakingWrapper } = deployment.staking;
+
+    return new FunctionAssertion<{}>(stakingWrapper.moveStake, {
+        after: async (
+            _beforeInfo,
+            _result,
+            from: StakeInfo,
+            to: StakeInfo,
+            amount: BigNumber,
+            txData: Partial<TxData>,
+        ) => {
+            logUtils.log(
+                `moveStake({status: ${StakeStatus[from.status]}, poolId: ${from.poolId} }, { status: ${
+                    StakeStatus[to.status]
+                }, poolId: ${to.poolId} }, ${amount})`,
+            );
+
+            const owner = txData.from as string;
+            const updatedPools = [];
+            if (from.status === StakeStatus.Undelegated) {
+                decrementNextEpochBalance(ownerStake[StakeStatus.Undelegated], amount);
+                decrementNextEpochBalance(globalStake[StakeStatus.Undelegated], amount);
+            } else if (from.status === StakeStatus.Delegated) {
+                _.defaults(ownerStake[StakeStatus.Delegated], {
+                    [from.poolId]: new StoredBalance(),
+                });
+                decrementNextEpochBalance(ownerStake[StakeStatus.Delegated][from.poolId], amount);
```

Comment:
Could be good to move this into a helper

Reply:
I can't think of a clean way to do this with the way that js passes arguments by reference 😕